### PR TITLE
fix: types: error out on decoding BlockMsg with extraneous data

### PR DESCRIPTION
**Pull request by AnomalRoil**
This fixes the OSS-fuzz finding [48208: lotus:fuzz_block_msg](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=48208).

The error being that invalid CBOR encoding of BlockMsg can lead to decoded values that do not re-serialize as is. Not necessarily a security issue, but making sure we're not decoding invalid representations shouldn't hurt and prevents us from having to worry about having multiple CBOR representation of the same thing (our CBOR decoder requires canonical CBOR).

## Related Issues
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=48208

## Proposed Changes
This reads the remaining bytes in the reader provided to `UnmarshalCBOR` and if it still contains data, errors out. 

## Additional Info
The oss-fuzz testcase boils down to:
`0x83F6808020`
which contains an extra byte after the data items: https://cbor.me/?bytes=83(F6-80-80((20

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

### CODE REVIEW

**Review by fridrik01 — APPROVED** [review]
LGTM, also nice tests!

File: chain/types/blockmsg_test.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	type args struct {
+		b []byte
+	}
+	tests := []struct {
+		name    string
+		data    []byte
+		want    *BlockMsg
+		wantErr bool
+	}{
+		{"decode empty BlockMsg with extra data at the end", []byte{0x83, 0xf6, 0x80, 0x80, 0x20}, new(BlockMsg), true},
```
Comment (ZenGround0) [inline chain/types/blockmsg_test.go:19]: nit: set `want` to `nil`

File: chain/types/blockmsg_test.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		data := tt.data
+		want := tt.want
+		wantErr := tt.wantErr
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := DecodeBlockMsg(data)
+			if wantErr {
+				assert.Errorf(t, err, "DecodeBlockMsg(%v)", data)
+				return
+			}
+			assert.NoErrorf(t, err, "DecodeBlockMsg(%v)", data)
```
Comment (ZenGround0) [inline chain/types/blockmsg_test.go:33]: nit: print data as `%x` for better readability

**Review by ZenGround0 — APPROVED** [review]

**Review by arajasek — APPROVED** [review]

**Comment by arajasek** [comment]
Thanks very much!

**Pull request merged**